### PR TITLE
Add mutual close

### PR DESCRIPTION
**Pull request by DariusParvin**
fill in todos for:
- `authorize_mutual_close`
- `mutual_close`

I tested it and it worked - the transaction went through and the channel status was set to `closed`.

Remaining todo I wasn't sure how to handle:
I created `MutualCloseError` so that `mutual_close` would return an error consistent with the other contract call operations. I then commented out the abort condition when getting `InvalidAuthorizationSignature` (close.rs L434) because I'm not sure how the error types work.

### CODE REVIEW

File: src/bin/customer/close.rs [hunk 1]
```diff
@@ -420,23 +420,35 @@ async fn mutual_close(
...
 
-    // Otherwise, close the dialectic channel...
+    // Call the mutual close entrypoint
+    let mutual_close_result = tezos_client
+        .mutual_close(
+            close_state.customer_balance(),
+            close_state.merchant_balance(),
+            &authorization_signature,
+        )
+        .await;
```
[the remark this answers is not in the document]
Comment (DariusParvin) [reply src/bin/customer/close.rs:453]: Is this sufficient for error handling?

```
    mutual_close_result.context(format!(
        "Failed to call mutual close for {}",
        close.label.clone()
    ))?;
    ```
It was there already but a bit further down. After making the change Marcella suggested above, It'll be immediately below the line you commented.

File: src/escrow/tezos.rs [hunk 2]
```diff
@@ -310,12 +312,52 @@ fn python_context() -> inline_python::Context {
...
+            customer_balance, merchant_balance
+        ):
+            // Merchant pytezos interface
+            merch_py = pytezos.using(key=merch_acc, shell=uri)
+            // Specify the structure and types of the fields going into the mutual close state.
+            ty = MichelsonType.match(michelson_to_micheline("pair (pair bls12_381_fr string) (pair address (pair mutez mutez))"))
+            // create the packed (serialized) version of the mutual close state, corresponding to the types above.
+            // legacy=True ensures pytezos will always serialize the data as in michelson rather than micheline.
+            packed = ty.from_python_object((channel_id, "zkChannels mutual close", contract_id, customer_balance, merchant_balance)).pack(legacy=True).hex()
+            authorization_signature = merch_py.key.sign(packed)
```
[the remark this answers is not in the document]
Comment (DariusParvin) [reply src/escrow/tezos.rs:333]: I  added comments at the top of the python code of where the documentation lives. In reality I had to ask the devs in the Tezos slack to figure out how to do a lot of it though.

File: src/bin/customer/close.rs [hunk 3]
```diff
@@ -420,23 +420,35 @@ async fn mutual_close(
...
+
+    // Close the dialectic channel...
     proceed!(in chan);
     chan.close();
```
[the remark this answers is not in the document]
Comment (DariusParvin) [reply src/bin/customer/close.rs]: Done! Thanks for explaining what those lines do too

File: src/escrow/tezos.rs [hunk 4]
```diff
@@ -1269,8 +1294,77 @@ impl TezosClient {
...
+            .map_err(AuthorizeMutualCloseError)
+        }
+    }
+
+    /// Verify the authorization signature provided by the merchant during the mutual close flow.
+    /// The signature must be a valid EdDSA signature over the tuple
+    /// `(contract id, "zkChannels mutual close", channel id, customer balance, merchant balance)`
+    ///
+    /// This is called by the customer.
+    #[allow(unused)]
```
[the remark this answers is not in the document]
Comment (DariusParvin) [reply src/escrow/tezos.rs]: removed it for the other functions we use as well: `authorize_mutual_close` and `mutual_close`

[the remark this answers is not in the document]
Comment (marsella) [reply src/escrow/tezos.rs]: nice :sunglasses:

File: src/bin/customer/close.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (yaymukund) [reply src/bin/customer/close.rs:453]: Can you collapse them into a single line?

[the remark this answers is not in the document]
Comment (DariusParvin) [reply src/bin/customer/close.rs]: Ah yes good point. Done!

**Comment by DariusParvin** [comment]
latest force push was to rebase

**Review by marsella — APPROVED** [review]
This looks good to me! You'll still need to fix the clippy errors before you merge. Clippy in particular won't show all your errors at once, it needs to be re-run after you fix stuff to make sure there aren't new things it discovers. The CI runs the most exhaustive command:
```
cargo clippy --all-features --all-targets -- -Dwarnings
```

**Comment by DariusParvin** [comment]
ah okay I didn't know about the clippy errors. I just fixed them, thanks @marsella!

**Pull request merged**